### PR TITLE
Apply link filters & apply constraints in filters #LMR-1478

### PR DESCRIPTION
**Pull request by kubedo8**

### CODE REVIEW

File: src/app/core/store/collections/collection.util.ts [hunk 1]
```diff
@@ -75,3 +76,8 @@ export function createAttributesMap(attributes: Attribute[]): Record<string, Att
     return attributesMap;
   }, {});
 }
+
+export function findAttributeConstraint(attributes: Attribute[], attributeId: string): Constraint {
+  const attribute = (attributes || []).find(attr => attr.id === attributeId);
+  return attributes && attribute.constraint;
```
Comment (livthomas) [inline src/app/core/store/collections/collection.util.ts]: There should be `attribute` and not `attributes` in the return statement.

File: src/app/core/store/documents/documents.filters.ts [hunk 2]
```diff
@@ -198,13 +209,22 @@ function filterDocumentsByStem(
 }
 
 function cleanStemForBaseCollection(stem: QueryStem, documents: DocumentModel[]): QueryStem {
-  return cleanStemForCollection(stem, documents, stem.collectionId);
+  return cleanStemForCollectionAndLink(stem, documents, stem.collectionId);
```
Comment (livthomas) [inline src/app/core/store/documents/documents.filters.ts]: I have seen the same method on the backend and I just wonder, what do these clean methods do?

File: src/app/shared/data-input/boolean/boolean-data-input.component.scss [hunk 3]
```diff
@@ -8,3 +8,8 @@ div.custom-control {
   height: 1.5rem;
   padding-left: 2rem !important;
 }
+
+.custom-control-input {
+  top: 0;
+  right: 0;
```
Comment (livthomas) [inline src/app/shared/data-input/boolean/boolean-data-input.component.scss]: Does it work correctly across all perspectives?

File: src/app/shared/data-input/data-input.component.ts [hunk 4]
```diff
@@ -51,12 +70,71 @@ export class DataInputComponent {
...
+  }
+
+  private getWidthOfInput(value: any): number {
+    if (this.constraint && this.constraint.type === ConstraintType.Boolean) {
+      return 34;
+    }
+
+    if (!this.tempElement) {
+      this.tempElement = this.createTempElement();
+      document.body.appendChild(this.tempElement);
```
Comment (livthomas) [inline src/app/shared/data-input/data-input.component.ts]: What is this temporary element good for?

File: src/app/core/store/collections/collection.util.ts (hunk 1 above)
Comment (kubedo8) [reply src/app/core/store/collections/collection.util.ts]: good catch!

File: src/app/core/store/documents/documents.filters.ts (hunk 2 above)
Comment (kubedo8) [reply src/app/core/store/documents/documents.filters.ts]: The idea is to split `QueryStem` into separate stems which contains only concrete collection (and its link) based `filters`, `linkFilters` and `documentIds`. Then I'm filtering documents in multiples stages from the first collection (called as "base collection") to the latest collection in `QueryStem`.

File: src/app/shared/data-input/data-input.component.ts (hunk 4 above)
Comment (kubedo8) [reply src/app/shared/data-input/data-input.component.ts]: It it used for computing `input` width when `resizeToContent` mode is enabled.

File: src/app/shared/data-input/boolean/boolean-data-input.component.scss (hunk 3 above)
Comment (kubedo8) [reply src/app/shared/data-input/boolean/boolean-data-input.component.scss]: Yes.

**Comment by kubedo8** [comment]
@livthomas It's now ready to merge.

**Pull request merged**